primitives/npos-elections/fuzzer/src/phragmen_pjr.rs

```diff
@@ -0,0 +1,118 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Fuzzing which ensures that running unbalanced sequential phragmen always produces a result
+//! which satisfies our PJR checker.
+//!
+//! ## Running a single iteration
+//!
+//! Honggfuzz shuts down each individual loop iteration after a configurable time limit.
+//! It can be helpful to run a single iteration on your hardware to help benchmark how long that time
+//! limit should reasonably be. Simply run the program without the `fuzzing` configuration to run a
+//! single iteration: `cargo run --bin phragmen_pjr`.
+//!
+//! ## Running
+//!
+//! Run with `HFUZZ_RUN_ARGS="-t 10" cargo hfuzz run phragmen_pjr`.
+//!
+//! Note the environment variable: by default, `cargo hfuzz` shuts down each iteration after 1 second
+//! of runtime. We significantly increase that to ensure that the fuzzing gets a chance to complete.
+//! Running a single iteration can help determine an appropriate value for this parameter.
+//!
+//! ## Debugging a panic
+//!
+//! Once a panic is found, it can be debugged with
+//! `HFUZZ_RUN_ARGS="-t 10" cargo hfuzz run-debug phragmen_pjr hfuzz_workspace/phragmen_pjr/*.fuzz`.
+//!
+
+#[cfg(fuzzing)]
+use honggfuzz::fuzz;
+
+#[cfg(not(fuzzing))]
+use structopt::StructOpt;
+
+mod common;
+use common::{generate_random_npos_inputs, to_range};
+use rand::{self, SeedableRng};
+use sp_npos_elections::{pjr_check_core, seq_phragmen_core, setup_inputs, standard_threshold};
+
+type AccountId = u64;
+
+const MIN_CANDIDATES: usize = 5;
+const MAX_CANDIDATES: usize = 200;
+const MIN_VOTERS: usize = 5;
+const MAX_VOTERS: usize = 500;
```

Comment:
```suggestion
const MIN_CANDIDATES: usize = 500;
const MAX_CANDIDATES: usize = 2000;
const MIN_VOTERS: usize = 1000;
const MAX_VOTERS: usize = 5000;
```

These are more realistic polkadot values, and I think it should not get super slow yet.

Reply:
Unfortunately, it appears to be pretty slow. On my personal computer, building and executing a single iteration took 159s in release mode. On my Parity Macbook (2019 edition), it took 182s. 

I haven't yet done any real benchmarking to discover whether the slowdown is in case generation or in the `seq_phragmen` algorithm, or both, but I think iterations measured in minutes are too slow for effective fuzzing.

What's clear is that at least one of those two stages is nonlinear in complexity. Cutting your suggested case sizes in half gives a timing of 19.2s on my personal machine, and 22.6s on the MBP. I think that's a better balance of realism and iteration time.

Reply:
9f928e3a1

Reply:
Okay no 159 is way too low. Make it something that is feasible on 1 machine. 

I was using the build-host with 128 cores and got a bit spoiled I think :D